src/lib.rs

```diff
@@ -100,82 +112,284 @@ pub async fn run_cargo_command(
+                            // was placed at `*.component.wasm`, so we use that to overwrite the
+                            // `*.wasm` file. In the process we also create a `*.module.wasm` for
+                            // debugging.
+                            //
+                            // Note that we remove files before renaming and such to ensure that
+                            // we're not accidentally updating the wrong hard link and such.
+                            let temporary_module =
+                                parent.join(basename).with_extension("module.wasm");
+                            let temporary_component =
+                                parent.join(basename).with_extension("component.wasm");
+
+                            drop(fs::remove_file(&temporary_module));
+                            fs::rename(wasm, &temporary_module)?;
+                            if !*fresh || !temporary_component.exists() {
+                                let path = &temporary_module;
+                                let bytes = &mut fs::read(path).with_context(|| {
+                                    format!(
+                                        "failed to read output module `{path}`",
+                                        path = path.display()
+                                    )
+                                })?;
+
+                                // If the compilation output is not a WebAssembly module, then do nothing
+                                // Note: due to the way cargo currently works on macOS, it will overwrite
+                                // a previously generated component on an up-to-date build.
+                                //
+                                // As a result, users on macOS will see a "creating component" message
+                                // even if the build is up-to-date.
+                                //
+                                // See: https://github.com/rust-lang/cargo/blob/99ad42deb4b0be0cdb062d333d5e63460a94c33c/crates/cargo-util/src/paths.rs#L542-L550
+                                if bytes.len() < 8 || bytes[0..4] != [0x0, b'a', b's', b'm'] {
+                                    bail!(
+                                        "expected `{path}` to be a WebAssembly module",
+                                        path = path.display()
+                                    );
+                                }
+                                // Check for the module header version
+                                if bytes[4..8] == [0x01, 0x00, 0x00, 0x00] {
+                                    if is_test {
+                                        // To prevent the test entrypoint to clash with default WASI
+                                        // command export, we will rename it to something random
+                                        replace_export(bytes);
```

Comment:
The test/bench command will not work if the user overrides the adapter that is not of type command. I think the best approach would be to actually modify whatever adapter is resolved to so that the `run` export points to the test entrypoint. Unfortunately, I could not figure out how to do it. I did this hack of replacing the export `run` so at least it will work in the case of a reactor project with wasi cli that has no override.